csvcubed/csvcubed/models/codelistconfig/code_list_config.py

```diff
@@ -0,0 +1,86 @@
+from typing import Optional, List, Tuple, Dict
+from dataclasses import dataclass, field
+from pathlib import Path
+
+from csvcubedmodels.dataclassbase import DataClassBase
+
+from csvcubed.models.cube.qb.catalog import CatalogMetadata
+from csvcubed.utils.json import load_json_document
+from csvcubed.models.cube.qb.components import NewQbConcept
+from csvcubed.readers.catalogmetadata.v1.catalog_metadata_reader import (
+    metadata_from_dict,
+)
+
+CODE_LIST_CONFIG_DEFAULT_URL = "https://purl.org/csv-cubed/code-list-config/v1.0"
+
+
+@dataclass
+class CodeListConfigSort(DataClassBase):
+    """Model for representing the sort object in code list config."""
+
+    by: str
+    method: str
+
+
+@dataclass
+class CodeListConfigConcept(DataClassBase):
+    """Model for representing a code list concept in code list config."""
+
+    label: str
+    notation: str
+    parent_notation: Optional[str] = field(default=None)
+    description: Optional[str] = field(default=None)
+    sort_order: Optional[int] = field(default=None)
+    same_as: Optional[str] = field(default=None)
+    children: List["CodeListConfigConcept"] = field(default_factory=list)
+
+
+@dataclass
+class CodeListConfig(DataClassBase):
+    """Model for representing a code list config."""
+
+    sort: Optional[CodeListConfigSort] = field(default=None)
+    concepts: List[CodeListConfigConcept] = field(default_factory=list)
+    schema: str = field(init=False, default=CODE_LIST_CONFIG_DEFAULT_URL)
+    metadata: CatalogMetadata = field(
+        init=False, default_factory=lambda: CatalogMetadata("Metadata")
+    )
+
+    @classmethod
+    def from_json_file(cls, file_path: Path) -> Tuple["CodeListConfig", Dict]:
+        """
+        Converts code list config json to `CodeListConfig`.
+        """
+        code_list_dict = load_json_document(file_path)
+        schema = code_list_dict.get("$schema", CODE_LIST_CONFIG_DEFAULT_URL)
+
+        code_list_config = cls.from_dict(code_list_dict)
+        code_list_config.schema = schema
+        code_list_config.metadata = metadata_from_dict(code_list_dict)
+
+        return (code_list_config, code_list_dict)
+
+    @property
+    def new_qb_concepts(self) -> list[NewQbConcept]:
+        """
+        Converts concepts of type CodeListConfigConcept to concepts of type NewQbConcept whilst maintaining the hierarchy.
+        """
+
+        new_qb_concepts: list[NewQbConcept] = []
+        if self.concepts:
+            concepts: list[CodeListConfigConcept] = list(self.concepts)
+
+            for concept in concepts:
+                new_qb_concepts.append(
+                    NewQbConcept(
+                        label=concept.label,
+                        code=concept.notation,
+                        parent_code=concept.parent_notation,
+                        sort_order=concept.sort_order,
```

Comment:
Are we applying the sort order as defined in the `sort` attribute? I'd say if the concept has a sort order defined directory on it, then use that, else go with the sorting defined in the `sort` attribute.

Reply:
@robons So we want to sort the concepts based on the `sort` object or `sort_order` in python? If so, I have the following questions:
1. Can the user define both the `sort` object and `sort_order` for all or some of the concepts?
2. If the answer to the above is _Yes_, what sort should be prioritised?
3. When sorting, should we sort all the concepts as a flat list or do we need to sort within each hierarchy if the code list contains hierarchical concepts?

Reply:
> 1. Can the user define both the sort object and sort_order for all or some of the concepts?
> 2. If the answer to the above is Yes, what sort should be prioritised?

As I say above, start out by applying the `sort` and then let the user override whatever that results in by specifying a `sort_order` manually. 

> 3. When sorting, should we sort all the concepts as a flat list or do we need to sort within each hierarchy if the code list contains hierarchical concepts?

Each set of child nodes within the hierarchy needs its own sort order (starting from 1 or 0 again).